main/src/ca/uwaterloo/flix/language/phase/optimizer/LambdaDrop.scala

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright 2025 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.optimizer
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.MonoAst.{Expr, empty}
+import ca.uwaterloo.flix.language.ast.shared.{BoundBy, Scope}
+import ca.uwaterloo.flix.language.ast.{MonoAst, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.dbg.AstPrinter.DebugMonoAst
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import scala.collection.mutable
+
+/**
+  * Rewrites recursive higher-order functions to non-recursive functions with
+  * a recursive local def where constant parameters become closure captured.
+  *
+  * A constant parameter is one that has the same value in every recursive call.
+  *
+  * Example:
+  *
+  * {{{
+  *   def map(f: a -> b, l: List[a]): List[b] = match l {
+  *       case Nil     => Nil
+  *       case x :: xs => f(x) :: map(f, xs)
+  *   }
+  * }}}
+  *
+  * Is rewritten to:
+  *
+  * {{{
+  *   def map(f: a -> b, l: List[a]): List[b] =
+  *       def map$Loop(l1) = match l1 {
+  *           case Nil      => Nil
+  *           case x :: xs  => f(x) :: map$Loop(xs)
+  *       };
+  *       map$Loop(l)
+  * }}}
+  */
+object LambdaDrop {
+
+  def run(root: MonoAst.Root)(implicit flix: Flix): MonoAst.Root = flix.phase("LambdaDrop") {
+    val newDefs = ParOps.parMapValues(root.defs)(visitDef)
+    root.copy(defs = newDefs)
+  }
+
+  private def visitDef(defn: MonoAst.Def)(implicit flix: Flix): MonoAst.Def = {
+    implicit val ctx: LocalContext = LocalContext.mk()
+    if (isRewritable(defn))
+      rewriteDefn(defn)
+    else
+      defn
+  }
+
+  private def isRewritable(defn: MonoAst.Def)(implicit ctx: LocalContext): Boolean = {
```

Comment:
I think a  doc comment that says we only LambdaDrop recursive higher-order functions.